include/urg_node/accessor.h

```diff
@@ -0,0 +1,95 @@
+/*
+ * accessor.h
+ *
+ *  Created on: 26/07/2022
+ *      Author: cribeiromendes
+ */
+
+#ifndef INCLUDE_URG_NODE_ACCESSOR_H_
+#define INCLUDE_URG_NODE_ACCESSOR_H_
+
+#include <sstream>
+#include <string>
+
+/**
+ * @brief Pattern taken from
+ * https://github.com/ros-drivers/um7/blob/466e6e05a5bba0cd0e2bd20c8fbb7101fc71e872/include/um7/registers.h
+ */
+
+/**
+ * @brief This class provides an accessor of fields.
+ */
+class Accessor_
+{
+public:
+  /**
+   * @brief Default C'tor
+   * @param buffer buffer where fields are accessed
+   * @param field_index Index of the field in the buffer
+   * @param field_width Size of the field in the buffer
+   */
+  Accessor_(const std::string* buffer, uint8_t field_index, uint8_t field_width) :
+    index(field_index),
+    width(field_width),
+    buffer_(buffer)
+  {
+  }
+
+  /**
+   * @brief Index of the field in the buffer
+   */
+  const uint8_t index;
+
+  /**
+   * @brief Size of the field in the buffer in bytes
+   */
+  const uint8_t width;
+
+protected:
+  /**
+   * @brief Pointer to buffer
+   */
+  const std::string* buffer_;
+};
+
+/**
+ * @brief Accessor for the field of type TField
+ * @tparam TField
+ */
+template <typename TField, size_t TIndex>
+class Accessor : public Accessor_
+{
+public:
+  /**
+   * @brief Default C'tor
+   * @param buffer buffer where fields are accessed
+   * @param field_index Index of the field in the buffer
+   * @param offset Custom offset that needs to be applied to the field
+   */
+  Accessor(const std::string* buffer, const uint8_t idx_offset = 0, const uint8_t param_offset = 0) :
+    Accessor_(buffer, TIndex + idx_offset, sizeof(TField)),
+    offset_(param_offset)
+  {
+  }
+
+  /**
+   * @brief Method to retrieve field from buffer
+   * @param field
+   */
+  void get(TField& field) const
+  {
+    std::stringstream ss;
+    ss << buffer_->substr(this->index, this->width);
+    ss >> std::hex >> field;
+    field += offset_;
+  }
+
+private:
+  /**
+   * @brief Custom offset that need to be added into the field
+   */
+  uint8_t offset_ { 0 };
```

Comment:
Should the offset be held by the base class as well? It doesn't seem like the offset functionality is specific to knowing the field type.

Along those same lines, would it make sense to have a template function access in the base class:
```
class Accessor_
{
public:
  template <typename TField>
  void get(TField& field) const
  {
    std::stringstream ss;
    ss << buffer_->substr(this->index, this->width);
    ss >> std::hex >> field;
    field += offset_;
  }
```

And then specialize it (shadow it?) in the derived class:
```
template <typename TField, size_t TIndex>
class Accessor : public Accessor_
{
public:
  void get(TField& field) const
  {
    Accessor_::get<TField>(field);
  }
```
I'm just trying to make the functionality split between `Accessor_` and `Accessor` make more sense. Why does the `Accessor_` class hold `index` and `width` if it never uses them? Why does `Accessor_` _not_ hold `offset_`?

---

Having said all of that, what is the reason for the base class at all? I didn't see the parameters being accessed polymorphically anywhere. Why not just have the template class version include all the functionality of the `Accessor_` class as well?

```
template <typename TField, size_t TIndex>
class Accessor : public Accessor_
{
public:
  /**
   * @brief Default C'tor
   * @param buffer buffer where fields are accessed
   * @param field_index Index of the field in the buffer
   * @param offset Custom offset that needs to be applied to the field
   */
  Accessor(const std::string* buffer, const uint8_t idx_offset = 0, const uint8_t param_offset = 0) :
    index_(buffer, TIndex + idx_offset),
    width_(sizeof(TField)),
    offset_(param_offset)
  {
  }

  /**
   * @brief Method to retrieve field from buffer
   * @param field
   */
  void get(TField& field) const
  {
    std::stringstream ss;
    ss << buffer_->substr(this->index, this->width);
    ss >> std::hex >> field;
    field += offset_;
  }

private:
  /**
   * @brief Index of the field in the buffer
   */
  const uint8_t index;

  /**
   * @brief Size of the field in the buffer in bytes
   */
  const uint8_t width;

  /**
   * @brief Custom offset that need to be added into the field
   */
  uint8_t offset_ { 0 };
};
```

Reply:
Initially, I thought that it would make sense to have the helper/handler structure hold an abstract accessor, but eventually, I did not find a reason to do it and instantiated it directly. I will remove this.